[JSC] Refine DFG AI type speculations for GetInternalField when called on ProxyObject

shvaikalesh opened the pull request:
#### ff7f706b1d907c0c1fe8b55d4a6459e3cd4496c9
<pre>
[JSC] Refine DFG AI type speculations for GetInternalField when called on ProxyObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=254271">https://bugs.webkit.org/show_bug.cgi?id=254271</a>
&lt;rdar://problem/107056889&gt;

Reviewed by Yusuke Suzuki.

This change refines type speculations for [[ProxyTarget]] and [[ProxyHandler]] internal fields
as their types are guaranteed by the ECMA-262, getting rid of SpecBytecodeTop, which enables downstream
optimizations in IC helpers for ProxyObject&apos;s &quot;has&quot; and &quot;get&quot; traps, progressing microbenchmarks by 6-10%.

microbenchmarks/proxy-get-miss-handler.js speedup probably slighly affects Speedometer2/Flight-TodoMVC.

This very optimization may be applied to other object types, including Promise.

                                    ToT                     patch

proxy-has-hit                 4.9520+-0.0588     ^      4.6705+-0.0609        ^ definitely 1.0603x faster
proxy-has-miss-handler        5.0585+-0.0674     ^      4.7937+-0.0778        ^ definitely 1.0552x faster
proxy-get-miss-handler       17.1412+-0.1817     ^     15.5256+-0.1962        ^ definitely 1.1041x faster

* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):

Canonical link: <a href="https://commits.webkit.org/262054@main">https://commits.webkit.org/262054@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/951c549adc44d80051d32817f9fd0c0c3d409c7f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/411 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/424 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/443 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/411 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/466 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/486 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/614 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/416 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/397 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/442 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/386 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/362 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/358 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/402 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/415 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/385 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/428 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/397 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/67 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/98 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/393 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/428 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/65 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -3487,6 +3487,17 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
         break;
 
     case GetInternalField:
+        if (node->child1()->shouldSpeculateProxyObject()) {
```
We should not use speculation in AI. Look into AbstractValue's type instead.

Constellation commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -3518,9 +3518,22 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
-    case GetInternalField:
+    case GetInternalField: {
+        AbstractValue child = forNode(node->child1());
+        if (!(child.m_type & ~SpecProxyObject)) {
```
And ensure that child.m_type is not clear.

Constellation reviewed (approved):
r=me with comment

Commit:
[JSC] Refine DFG AI type speculations for GetInternalField when called on ProxyObject
https://bugs.webkit.org/show_bug.cgi?id=254271
<rdar://problem/107056889>

Reviewed by Yusuke Suzuki.

This change refines type speculations for [[ProxyTarget]] and [[ProxyHandler]] internal fields
as their types are guaranteed by the ECMA-262, getting rid of SpecBytecodeTop, which enables downstream
optimizations in IC helpers for ProxyObject's "has" and "get" traps, progressing microbenchmarks by 6-10%.

microbenchmarks/proxy-get-miss-handler.js speedup probably slighly affects Speedometer2/Flight-TodoMVC.

This very optimization may be applied to other object types, including Promise.

                                    ToT                     patch

proxy-has-hit                 4.9520+-0.0588     ^      4.6705+-0.0609        ^ definitely 1.0603x faster
proxy-has-miss-handler        5.0585+-0.0674     ^      4.7937+-0.0778        ^ definitely 1.0552x faster
proxy-get-miss-handler       17.1412+-0.1817     ^     15.5256+-0.1962        ^ definitely 1.1041x faster

* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):

Canonical link: https://commits.webkit.org/262054@main
